### PR TITLE
Refactor `cross_entropy` using `log_softmax` and `nll_loss` references

### PR DESCRIPTION
**Pull request by rdspring1**
This PR refactors `cross_entropy` to use `log_softmax` and `nll_loss`. It also removes `cross_entopy_backward` because we'll use `log_softmax_backward` and `nll_loss_backward` directly.

It enables the label smoothing and probability target test cases for the cross-entropy opinfo test. 

Fixes Issue https://github.com/Lightning-AI/lit-thunder-LEGACY/issues/704

### CODE REVIEW

File: thunder/tests/opinfos.py [hunk 1]
```diff
@@ -7805,6 +7910,11 @@ def cross_entropy_sample_generator(op, device, dtype, requires_grad, **kwargs):
                 datatypes.bfloat16,
             ),
         ),
+        # TODO FIXME -- These tests are hitting an odd issue where real torch tensors are being passed to nll_loss
```
Comment (rdspring1) [inline thunder/tests/opinfos.py:7913]: This skip is inherited from `nll_loss` because we're using it in the cross-entropy decomposition.

**Comment by lantiga** [comment]
@rdspring1 the failure is due to #410 renaming `ddp_type` to `distparallel_type`, it should be a superquick fix

**Comment by lantiga** [comment]
it's been resolved, @rdspring1 can you merge `main`? thanks!

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -3590,253 +3576,191 @@ def cross_entropy(
         lambda: f"Deprecated size_average={size_average} and reduce={reduce} is not supported!",
     )
 
+    _cross_entropy_input_checks(a, target, weight, ignore_index, reduction, label_smoothing)
+
+    # channels dimension is either the first one if no batch dim present (i.e. a.shape[0]),
+    # or right next to it (i.e. a.shape[1]).
+    channels_dim = 1 if a.ndim >= 2 else 0
```
Comment (lantiga) [inline thunder/torch/__init__.py:3583]: `channels` is very particular to image data, can we name this differently, like `softmax_dim`, or `reduction_dim`?

File: thunder/torch/__init__.py [hunk 3]
```diff
@@ -3590,253 +3576,191 @@ def cross_entropy(
...
-            return zeros(output_shape, device=a.device, dtype=a.dtype)
-        elif reduction == "mean":
-            fill_value = float("nan")
-        elif reduction == "sum":
-            fill_value = 0.0
-        else:
-            raise ValueError(f"Reduction argument {reduction} to cross_entropy is not supported")
+    # channels dimension is either the first one if no batch dim present (i.e. a.shape[0]),
+    # or right next to it (i.e. a.shape[1]).
+    channels_dim = 1 if a.ndim >= 2 else 0
```
Comment (lantiga) [inline thunder/torch/__init__.py:3637]: same comment as above

**Review by lantiga — COMMENTED** [review]
LGTM, just a minor comment from me, but I'd like a review from an op expert before we merge

File: thunder/torch/__init__.py (hunk 2 above)
Comment (t-vi) [reply thunder/torch/__init__.py:3583]: I think "class" is the word we're looking for per the [PyTorch docs for `cross_entropy`](https://pytorch.org/docs/master/generated/torch.nn.functional.cross_entropy.html).

**Review by t-vi — CHANGES_REQUESTED** [review]
Thank you @rdspring1 @lantiga 
Once we change the misnomer and write "class" instead of "channel", this seems good to merge.

**Review by t-vi — APPROVED** [review]
On further thought, let's merge and then fix the tidbits.

File: thunder/core/transforms.py [hunk 4]
```diff
@@ -1091,7 +1091,10 @@ def _div_prim_grad(a: Number | TensorProxy, b: Number | TensorProxy, /) -> Numbe
 
 # Comparison operators -- these create no grad associations
 register_grad(pids.EQ, prims.eq)
+register_grad(pids.NE, prims.ne)
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply thunder/core/transforms.py:1094]: Yes let's remove this.

I keep adding them myself as well :laughing:

File: thunder/tests/opinfos.py [hunk 5]
```diff
@@ -7776,21 +7777,125 @@ def cross_entropy_sample_generator(op, device, dtype, requires_grad, **kwargs):
...
+    yield (
+        SampleInput(valid_input, valid_prob_target, ignore_index=5),
+        RuntimeError,
+        "ignore_index argument is not supported when target contains class probabilities.",
+    )
+
+
 cross_entropy_opinfo = OpInfo(
     ltorch.cross_entropy,
-    supports_grad=True,
```
Comment (jjsjann123) [inline thunder/tests/opinfos.py:7788]: supports_grad is switched to false, this means we are not testing all the configuration for backward?

Does `supports_grad` require the operation to be an explicit autograd op?

File: thunder/tests/opinfos.py [hunk 6]
```diff
@@ -7776,21 +7777,125 @@ def cross_entropy_sample_generator(op, device, dtype, requires_grad, **kwargs):
...
     sample_input_generator=cross_entropy_sample_generator,
     reference_input_generator=cross_entropy_reference_generator,
+    error_input_generator=cross_entropy_error_generator,
     torch_reference=torch.nn.functional.cross_entropy,
     dtypes=(datatypes.floating,),
     test_directives=(
+        # take_along_axis is disabled with nvfuser, which this operator relies on.
+        DecorateInfo(
+            pytest.mark.skip,
+            executors=("nvfuser",),
```
Comment (jjsjann123) [inline thunder/tests/opinfos.py:7897]: qq: this shouldn't be a function thing right? i.e. nvfuser can reject part of the fusion, but running it with nvfuser executor shouldn't break IIUC.

File: thunder/torch/__init__.py [hunk 7]
```diff
@@ -3590,253 +3576,191 @@ def cross_entropy(
...
+
+    # NOTE This short-circuit is subject to change and is placed ahead of other input checks to match PyTorch behavior.
+    # The expected behavior when the target and input have zero elements:
+    #   reduction = 'none' --- tensor([], shape) or tensor(0.)
+    #   reduction = 'sum'  --- tensor(0.)
+    #   reduction = 'mean' --- tensor(nan)
+    # Mean reduction on empty tensors produces NaN.
+    if a.numel() == 0:
+        if reduction == "none":
```
Comment (jjsjann123) [inline thunder/torch/__init__.py:3592]: Note to myself. Another justification for op-based constraint. `reduction` string needs to be baked-in and properly checked in prologue trace.

File: thunder/torch/__init__.py [hunk 8]
```diff
@@ -3590,253 +3576,191 @@ def cross_entropy(
...
+            output_shape = list(a.shape)
+            output_shape.pop(channels_dim)
+            return full(output_shape, 0.0, device=a.device, dtype=a.dtype)
+        elif reduction == "sum":
+            return full(result_shape := [], fill_value := 0.0, device=a.device, dtype=a.dtype)
+        elif reduction == "mean":
+            return full(result_shape := [], fill_value := float("nan"), device=a.device, dtype=a.dtype)
+
+    if a.shape == target.shape:
```
Comment (jjsjann123) [inline thunder/torch/__init__.py:3601]: Note to myself: shape based conditional! This is gonna be bad. linking #262
